libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/AlternativeRouteProgressDataProvider.kt

```diff
@@ -0,0 +1,37 @@
+package com.mapbox.navigation.core.routealternatives
+
+import com.mapbox.navigation.core.RouteProgressData
+
+internal object AlternativeRouteProgressDataProvider {
+
+    fun getRouteProgressData(
+        primaryRouteProgressData: RouteProgressData,
+        alternativeMetadata: AlternativeRouteMetadata
+    ): RouteProgressData {
+        val legIndex: Int
+        val routeGeometryIndex: Int
+        val legGeometryIndex: Int?
+        val primaryFork = alternativeMetadata.forkIntersectionOfPrimaryRoute
+        val alternativeFork = alternativeMetadata.forkIntersectionOfAlternativeRoute
+        val isForkPointAheadOnPrimaryRoute =
+            primaryRouteProgressData.routeGeometryIndex < primaryFork.geometryIndexInRoute
+        if (isForkPointAheadOnPrimaryRoute) {
+            val legIndexDiff = primaryFork.legIndex - alternativeFork.legIndex
+            legIndex = primaryRouteProgressData.legIndex - legIndexDiff
+            val routeGeometryIndexDiff =
```

Comment:
As I know, even if 2 routes follow the same route, they could have a different points in their geometry. I remember that Directions API inserts additional points if traffic between two points changes. I don't know which range of error in index we should expect and if it's critical. Maybe it is, maybe it's not. Anyway current solution makes think better, so maybe it's more for followup PR.

I can propose following alternatives which shouldn't be affected by it:

Make the NN track current alternative geometry index.
Track it by ourselves: find nearest point in geometry to the current map matched position. As optimisation we can start search from routeGeometryIndexDiff as it should be very close.
https://github.com/orgs/mapbox/teams/navnative , any suggestions from your side?

Reply:
> Make the NN track current alternative geometry index.

That would be perfect, of course. :)

> Track it by ourselves: find nearest point in geometry to the current map matched position. As optimisation we can start search from routeGeometryIndexDiff as it should be very close.

I'm not sure it's a very good idea to implement this platform-side.

Reply:
> As I know, even if 2 routes follow the same roads, they could have a different points in their geometry. I remember that Directions API inserts additional points if traffic between two points varies. I don't know which range of error in index we should expect and if it's critical. Maybe it is, maybe it's not. Anyway current solution makes things better, so maybe this suggestion is more for a followup PR.

Where would the fork be in this case? If the routes follow the same roads but the geometries are different? @mskurydin do you know?

Reply:
I'm not sure about routes that have actually different geometry, but there can be routes that have different number of geometry points.

For example, if there's an incident that appeared later on in the trip, the alternative route might have it while the primary route won't. So even though the actual geometry doesn't change, the introduction of an incident will make the alternative route have one more point than the primary route (the point where incident happened). This is a universal risk that we're carrying and a major problem with the refresh feature in itself.

I think that's just the risk will have to bear with for now (and make sure that the code doesn't crash in these situations).

Reply:
> the introduction of an incident will make the alternative route have one more point than the primary route 

As far as I remember we Directions API also adds points for traffic when it needs more granularity, doesn't it?

Reply:
Yes, I believe so.